### PR TITLE
Restructure NXP MCUX USB driver

### PR DESCRIPTION
**Pull request by mmahadevan108**
1. Move the defines from usb_dc_mcux.h to usb_device_config.h and fsl_os_abstraction.h. These headers are used by the NXP SDK USB driver. usb_dc_mcux.h header file is not longer needed and hence deleted.
2. Delete the Zephyr implementation of the usb_device_struct driver and use the one implemented inside the SDK USB driver. This requires updating references to usb_device_struct inside the NXP MCUX Zephyr USB driver
3. Move defines and structures used by the Zephyr USB driver out of the header file that is included by the SDK and into the driver.
4. Update west.yml to update the NXP HAL to get the updated SDK USB driver.
5. Fix an issue with the RT10xx USB driver in High Speed mode.

### CODE REVIEW

File: modules/hal_nxp/fsl_os_abstraction.h [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __FSL_OS_ABSTRACTION__
+#define __FSL_OS_ABSTRACTION__
+
+#include <zephyr/irq.h>
+
+/* enter critical macros */
+#define OSA_SR_ALLOC() int osa_current_sr
```
[the remark this answers is not in the document]
Comment (mmahadevan108) [reply modules/hal_nxp/fsl_os_abstraction.h:13]: This was moved into Zephyr per an earlier request to move all Zephyr specific code into the Zephyr tree.

**Comment by hakehuang** [comment]
@mmahadevan108  I rebase the code, and run test on RT11xx and RT10xx for usb cases which is OK.

File: dts/bindings/usb/nxp,mcux-usbd.yaml [hunk 2]
```diff
@@ -14,3 +14,25 @@ properties:
 
     interrupts:
       required: true
+
+    usb-controller-index:
```
[the remark this answers is not in the document]
Comment (mmahadevan108) [reply dts/bindings/usb/nxp,mcux-usbd.yaml:18]: Done, thank you for the suggestion.

**Review by danieldegrasse — DISMISSED** [review]

**Review by danieldegrasse — DISMISSED** [review]

File: drivers/usb/device/usb_dc_mcux.c [hunk 3]
```diff
@@ -34,6 +38,12 @@ static void usb_isr_handler(void);
 #define SETUP_DATA_STAGE_DONE	(0)
 #define SETUP_DATA_STAGE_IN	(1)
 #define SETUP_DATA_STAGE_OUT	(2)
+#define USBD_MCUX_THREAD_STACK_SIZE	(1024)
+
+/* endpoint related macros */
+#define EP0_MAX_PACKET_SIZE	(64)
```
Comment (jfischer-no) [inline drivers/usb/device/usb_dc_mcux.c:43]: Please use USB_CONTROL_EP_MPS from include/zephyr/usb/usb_ch9.h. btw () do not make any sense here.

File: drivers/usb/device/usb_dc_mcux.c [hunk 4]
```diff
@@ -34,6 +38,12 @@ static void usb_isr_handler(void);
...
+#define EP0_OUT			(0)
```
Comment (jfischer-no) [inline drivers/usb/device/usb_dc_mcux.c:44]: Use USB_CONTROL_EP_IN  and USB_CONTROL_EP_IN instead.

File: drivers/usb/device/usb_dc_mcux.c [hunk 5]
```diff
@@ -74,8 +84,31 @@ K_HEAP_DEFINE_NOCACHE(ep_buf_pool, 1024 * EP_BUF_NUMOF_BLOCKS);
...
+	struct k_mem_block block;
+	usb_dc_ep_callback callback;
+	uint16_t ep_mps;
+	uint8_t ep_type;
+	uint8_t ep_enabled : 1;
+	uint8_t ep_occupied : 1;
+};
+
+struct usb_dc_state {
+	usb_device_struct_t usb_dev_struct;
```
Comment (jfischer-no) [inline drivers/usb/device/usb_dc_mcux.c:97]: Just `dev_struct`

File: drivers/usb/device/usb_dc_mcux.c [hunk 6]
```diff
@@ -74,8 +84,31 @@ K_HEAP_DEFINE_NOCACHE(ep_buf_pool, 1024 * EP_BUF_NUMOF_BLOCKS);
...
+	/* Controller handle */
+	usb_dc_status_callback status_callback;
```
Comment (jfischer-no) [inline drivers/usb/device/usb_dc_mcux.c:99]: callback is usually just cb, status_cb

File: drivers/usb/device/usb_dc_mcux.c [hunk 7]
```diff
@@ -74,8 +84,31 @@ K_HEAP_DEFINE_NOCACHE(ep_buf_pool, 1024 * EP_BUF_NUMOF_BLOCKS);
...
+	struct usb_ep_ctrl_data *eps;
+	bool attached;
+	uint8_t setupDataStage;
```
Comment (jfischer-no) [inline drivers/usb/device/usb_dc_mcux.c:102]: Style :arrow_up:

File: drivers/usb/device/usb_dc_mcux.c [hunk 8]
```diff
@@ -74,8 +84,31 @@ K_HEAP_DEFINE_NOCACHE(ep_buf_pool, 1024 * EP_BUF_NUMOF_BLOCKS);
...
+
+	K_KERNEL_STACK_MEMBER(thread_stack, USBD_MCUX_THREAD_STACK_SIZE);
+
+	struct k_thread thread;
+};
+
 static struct usb_ep_ctrl_data s_ep_ctrl[NUM_OF_EP_MAX];
-static struct usb_device_struct dev_data;
+static struct usb_dc_state dev_data;
```
Comment (jfischer-no) [inline drivers/usb/device/usb_dc_mcux.c:110]: This is still confusing, should probably be renamed to `static struct usb_dc_state dev_state`.

File: drivers/usb/device/usb_dc_mcux.c [hunk 9]
```diff
@@ -729,20 +778,20 @@ static void handle_transfer_msg(usb_device_callback_message_struct_t *cb_msg)
 	} else {
 		/* IN TOKEN */
 		if (USB_EP_DIR_IS_IN(ep)) {
-			if ((dev_data.address != 0) && (ep_abs_idx == 1)) {
+			if ((dev_data.usb_dev_struct.deviceAddress != 0) && (ep_abs_idx == 1)) {
```
Comment (jfischer-no) [inline drivers/usb/device/usb_dc_mcux.c:780]: Please move this block to handle_transfer_in() function, NACK for a code block like this that goes over 80 characters.

File: modules/hal_nxp/usb/usb_device_config.h [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+/******************************************************************************
+ * Definitions
+ *****************************************************************************/
+/* EHCI instance count */
+#ifdef CONFIG_USB_DC_NXP_EHCI
+#define USB_DEVICE_CONFIG_EHCI (1U)
+/* How many the DTD are supported. */
+#define USB_DEVICE_CONFIG_EHCI_MAX_DTD (16U)
+/* Whether device is self power. 1U supported, 0U not supported */
+#define USB_DEVICE_CONFIG_SELF_POWER (1U)
```
Comment (jfischer-no) [inline modules/hal_nxp/usb/usb_device_config.h:22]: How is it used? There is the same definition in the line 42.

**Review by jfischer-no — CHANGES_REQUESTED** [review]

**Review by hakehuang — DISMISSED** [review]

File: dts/arm/nxp/nxp_rt11xx.dtsi [hunk 11]
```diff
@@ -730,7 +729,6 @@
 			interrupt-names = "usb_otg";
 			clocks = <&xtal>;
 			num-bidir-endpoints = <8>;
-			maximum-speed = "full-speed";
```
[the remark this answers is not in the document]
Comment (mmahadevan108) [reply dts/arm/nxp/nxp_rt11xx.dtsi:732]: My removing this we are not limiting to full-speed.

File: modules/hal_nxp/usb/usb_device_config.h (hunk 10 above)
Comment (mmahadevan108) [reply modules/hal_nxp/usb/usb_device_config.h]: I have deleted the duplicate define. This is used in the SDK driver.

File: drivers/usb/device/usb_dc_mcux.c (hunk 3 above)
Comment (mmahadevan108) [reply drivers/usb/device/usb_dc_mcux.c]: Thanks for pointing this, I have made this change.

File: drivers/usb/device/usb_dc_mcux.c (hunk 4 above)
Comment (mmahadevan108) [reply drivers/usb/device/usb_dc_mcux.c]: Thanks for pointing this, I have made the change.

File: drivers/usb/device/usb_dc_mcux.c (hunk 9 above)
Comment (mmahadevan108) [reply drivers/usb/device/usb_dc_mcux.c:778]: I am not clear on this comment. This line is setting the same field inside the SDK data structure, the field name has changed with the update to the latest SDK. This passes the compliance check.

File: drivers/usb/device/usb_dc_mcux.c (hunk 8 above)
Comment (mmahadevan108) [reply drivers/usb/device/usb_dc_mcux.c:105]: Changed.

File: drivers/usb/device/usb_dc_mcux.c (hunk 6 above)
Comment (mmahadevan108) [reply drivers/usb/device/usb_dc_mcux.c:94]: CHanged

File: drivers/usb/device/usb_dc_mcux.c (hunk 5 above)
Comment (mmahadevan108) [reply drivers/usb/device/usb_dc_mcux.c:92]: Done

File: drivers/usb/device/usb_dc_mcux.c (hunk 7 above)
Comment (mmahadevan108) [reply drivers/usb/device/usb_dc_mcux.c:97]: Changed.

**Review by jfischer-no — APPROVED** [review]

**Review by EmilioCBen — APPROVED** [review]

**Pull request merged**